File: XRViewer/WebController/BarView.m
#import "BarView.h"
#import "OverlayHeader.h"

@interface URLTextField : UITextField
@end

@interface BarView () <UITextFieldDelegate>

@property (weak, nonatomic) IBOutlet URLTextField *urlField;
@property (weak, nonatomic) IBOutlet UIButton *backBtn;
@property (weak, nonatomic) IBOutlet UIButton *forwardBtn;
@property (nonatomic, weak) UIButton *reloadBtn;
@property (nonatomic, weak) UIButton *cancelBtn;
@property (nonatomic, weak) UIActivityIndicatorView *ai;
@property (nonatomic, weak) IBOutlet NSLayoutConstraint *rightConstraint;
@end

@implementation URLTextField

- (CGRect)leftViewRectForBounds:(CGRect)bounds
{
    return CGRectMake(0, 0, URL_FIELD_HEIGHT, URL_FIELD_HEIGHT);
}

@end


@implementation BarView

- (void)awakeFromNib
{
    [super awakeFromNib];
    
    [self setup];
}

- (NSString *)urlFieldText
{
    return [[self urlField] text];
}

- (void)startLoading:(NSString *)url
{
    [[self ai] startAnimating];
    [[self cancelBtn] setHidden:NO];
    [[self reloadBtn] setHidden:YES];
    [[self urlField] setText:url];
}

- (void)finishLoading:(NSString *)url
{
    [[self ai] stopAnimating];
    [[self cancelBtn] setHidden:YES];
    [[self reloadBtn] setHidden:NO];
}

- (void)setBackEnabled:(BOOL)enabled
{
    [[self backBtn] setEnabled:enabled];
}

- (void)setForwardEnabled:(BOOL)enabled
{
    [[self forwardBtn] setEnabled:enabled];
}

- (void)layout
{
    [self setupBarConstraint];
    
    [self layoutIfNeeded];
}

- (void)setup
{
    [[self backBtn] setImage:[UIImage imageNamed:@"back"] forState:UIControlStateDisabled];
    [[self forwardBtn] setImage:[UIImage imageNamed:@"forward"] forState:UIControlStateDisabled];
    [[self backBtn] setEnabled:NO];
    [[self forwardBtn] setEnabled:NO];
    
    [[self urlField] setDelegate:self];
    
    UIActivityIndicatorView *i = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
    [[self urlField] setLeftView:i];
    [[self urlField] setLeftViewMode:UITextFieldViewModeUnlessEditing];
    [i hidesWhenStopped];
    [self setAi:i];
    
    [[self urlField] setClearButtonMode:UITextFieldViewModeWhileEditing];
    [[self urlField] setReturnKeyType:UIReturnKeyGo];
    [[self urlField] setKeyboardType:UIKeyboardTypeURL];
    [[self urlField] setAutocorrectionType:UITextAutocorrectionTypeNo];
    
    [[self urlField] setTextContentType:UITextContentTypeURL];
    [[self urlField] setPlaceholder:@"Search or enter website name"];
    [[[self urlField] layer] setCornerRadius:URL_FIELD_HEIGHT / 4];
    [[self urlField] setTextAlignment:NSTextAlignmentCenter];
    
    UIButton *reloadButton = [UIButton buttonWithType:UIButtonTypeCustom];
    [reloadButton setImage:[UIImage imageNamed:@"reload"] forState:UIControlStateNormal];
    [reloadButton addTarget:self action:@selector(reloadAction:) forControlEvents:UIControlEventTouchDown];
    [reloadButton setFrame:CGRectMake(0, 0, URL_FIELD_HEIGHT, URL_FIELD_HEIGHT)];
    [reloadButton setHidden:NO];
    
    UIButton *cancelButton = [UIButton buttonWithType:UIButtonTypeCustom];
    [cancelButton setImage:[UIImage imageNamed:@"cancel"] forState:UIControlStateNormal];
    [cancelButton addTarget:self action:@selector(cancelAction:) forControlEvents:UIControlEventTouchDown];
    [cancelButton setFrame:CGRectMake(0, 0, URL_FIELD_HEIGHT, URL_FIELD_HEIGHT)];
    [cancelButton setHidden:YES];
    
    UIView *rightView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, URL_FIELD_HEIGHT, URL_FIELD_HEIGHT)];
    [rightView addSubview:reloadButton];
    [rightView addSubview:cancelButton];
    [self setCancelBtn:cancelButton];
    [self setReloadBtn:reloadButton];
    
    [[self urlField] setRightView:rightView];
    [[self urlField] setRightViewMode:UITextFieldViewModeUnlessEditing];
    
    [self setupBarConstraint];
}

- (void)setupBarConstraint
{
    BOOL isPortrait = UIInterfaceOrientationIsPortrait([[UIApplication sharedApplication] statusBarOrientation]);
    BOOL isntIpad = [[UIDevice currentDevice] userInterfaceIdiom] != UIUserInterfaceIdiomPad;
    
    [[self rightConstraint] setConstant:(isPortrait && isntIpad ? 27.5 : 127.5)];
}

- (IBAction)backAction:(id)sender
{
    DDLogDebug(@"backAction");
    
    [[self urlField] resignFirstResponder];
    
    if ([self backActionBlock])
    {
        [self backActionBlock](self);
    }
}

- (IBAction)forwardAction:(id)sender
{
    DDLogDebug(@"forwardAction");
    
    [[self urlField] resignFirstResponder];
    
    if ([self forwardActionBlock])
    {
        [self forwardActionBlock](self);
    }
}

- (IBAction)reloadAction:(id)sender
{
    DDLogDebug(@"reloadAction");
    
    [[self urlField] resignFirstResponder];
    
    if ([self reloadActionBlock])
    {
        [self reloadActionBlock](self);
    }
}

- (IBAction)cancelAction:(id)sender
{
    DDLogDebug(@"cancelAction");
    
    [[self urlField] resignFirstResponder];
    
    if ([self cancelActionBlock])
    {
        [self cancelActionBlock](self);
    }
}

#pragma mark UITextField Delegate

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    [textField resignFirstResponder];
    
    if ([self goActionBlock])
    {
        [self goActionBlock]([textField text]);
    }
    
    return YES;
}

#pragma mark UIView

- (nullable UIView *)hitTest:(CGPoint)point withEvent:(nullable UIEvent *)event
{
    CGFloat minXPos = CGRectGetMaxX([[self backBtn] frame]);
    CGFloat maxXPos = CGRectGetMinX([[self forwardBtn] frame]);
    
    CGFloat increaseValue = (maxXPos - minXPos) / 2;
    
    CGRect icreasedBackRect = CGRectMake([[self backBtn] frame].origin.x - increaseValue,
                                         [[self backBtn] frame].origin.y - increaseValue,
                                         [[self backBtn] frame].size.width + increaseValue * 2,
                                         [[self backBtn] frame].size.height + increaseValue * 2);
    
    CGRect icreasedForwardRect = CGRectMake([[self forwardBtn] frame].origin.x - increaseValue,
                                            [[self forwardBtn] frame].origin.y - increaseValue,
                                            [[self forwardBtn] frame].size.width + increaseValue * 2,
                                            [[self forwardBtn] frame].size.height + increaseValue * 2);
    
    if (CGRectContainsPoint(icreasedBackRect, point))
    {
        return [self backBtn];
    }
    
    if (CGRectContainsPoint(icreasedForwardRect, point))
    {
        return [self forwardBtn];
    }
    
    return [super hitTest:point withEvent:event];
}

@end

File: XRViewer/WebController/BarView.swift
import UIKit
import XCGLogger

let URL_FIELD_HEIGHT = 29

class BarView: UIView, UITextFieldDelegate {
    
    // MARK: - Properties & Outlets
    
    var backActionBlock: ((Any?) -> Void)?
    var forwardActionBlock: ((Any?) -> Void)?
    var homeActionBlock: ((Any?) -> Void)?
    var reloadActionBlock: ((Any?) -> Void)?
    var cancelActionBlock: ((Any?) -> Void)?
    var showPermissionsActionBlock: ((Any?) -> Void)?
    var goActionBlock: ((String?) -> Void)?
    var debugButtonToggledAction: ((Bool) -> Void)?
    var settingsActionBlock: (() -> Void)?
    var restartTrackingActionBlock: (() -> Void)?
    var switchCameraActionBlock: (() -> Void)?

    @IBOutlet weak var urlField: URLTextField!
    @IBOutlet private weak var backBtn: UIButton!
    @IBOutlet private weak var forwardBtn: UIButton!
    @IBOutlet private weak var homeBtn: UIButton!
    @IBOutlet private weak var debugBtn: UIButton!
    @IBOutlet private weak var settingsBtn: UIButton!
    private weak var reloadBtn: UIButton?
    private weak var cancelBtn: UIButton?
    weak var permissionLevelButton: ActivityIndicatorButton?
    @IBOutlet private weak var restartTrackingBtn: UIButton!
    @IBOutlet private weak var switchCameraBtn: UIButton!

    // MARK: - View Lifecycle
    
    override func awakeFromNib() {
        super.awakeFromNib()

        setup()
    }

    func setup() {
        backBtn.setImage(UIImage(named: "back"), for: .disabled)
        forwardBtn.setImage(UIImage(named: "forward"), for: .disabled)
        backBtn.isEnabled = false
        forwardBtn.isEnabled = false

        urlField.delegate = self

        let permissionButton = ActivityIndicatorButton(type: .custom)
        permissionButton.setImage(nil, for: .normal)
        permissionButton.addTarget(self, action: #selector(BarView.showPermissionsAction(_:)), for: .touchUpInside)
        permissionButton.frame = CGRect(x: 0, y: 0, width: CGFloat(URL_FIELD_HEIGHT), height: CGFloat(URL_FIELD_HEIGHT))
        permissionButton.imageEdgeInsets = UIEdgeInsets(top: 0, left: 5, bottom: 0, right: 5)
        permissionButton.isEnabled = false
        permissionLevelButton = permissionButton
        permissionLevelButton?.heightAnchor.constraint(equalToConstant: CGFloat(URL_FIELD_HEIGHT)).isActive = true
        permissionLevelButton?.widthAnchor.constraint(equalToConstant: CGFloat(URL_FIELD_HEIGHT)).isActive = true
        permissionLevelButton?.imageView?.contentMode = .scaleAspectFit
        urlField.leftView = permissionLevelButton
        urlField.leftViewMode = .unlessEditing

        urlField.clearButtonMode = .whileEditing
        urlField.returnKeyType = .go

        urlField.textContentType = .URL
        urlField.placeholder = "Search or enter website name"
        urlField.layer.cornerRadius = CGFloat(URL_FIELD_HEIGHT / 4)
        urlField.textAlignment = .center

        let reloadButton = UIButton(type: .custom)
        reloadButton.setImage(UIImage(named: "reload"), for: .normal)
        reloadButton.addTarget(self, action: #selector(BarView.reloadAction(_:)), for: .touchUpInside)
        reloadButton.frame = CGRect(x: 0, y: 0, width: CGFloat(URL_FIELD_HEIGHT), height: CGFloat(URL_FIELD_HEIGHT))
        reloadButton.isHidden = false

        let cancelButton = UIButton(type: .custom)
        cancelButton.setImage(UIImage(named: "cancel"), for: .normal)
        cancelButton.addTarget(self, action: #selector(BarView.cancelAction(_:)), for: .touchUpInside)
        cancelButton.frame = CGRect(x: 0, y: 0, width: CGFloat(URL_FIELD_HEIGHT), height: CGFloat(URL_FIELD_HEIGHT))
        cancelButton.isHidden = true

        let rightView = UIView(frame: CGRect(x: 0, y: 0, width: CGFloat(URL_FIELD_HEIGHT), height: CGFloat(URL_FIELD_HEIGHT)))
        rightView.addSubview(reloadButton)
        rightView.addSubview(cancelButton)
        self.cancelBtn = cancelButton
        self.reloadBtn = reloadButton

        urlField.rightView = rightView
        urlField.rightViewMode = .unlessEditing

        debugBtn.setImage(UIImage(named: "debugOff"), for: .normal)
        debugBtn.setImage(UIImage(named: "debugOn"), for: .selected)

        restartTrackingBtn.setImage(UIImage(named: "streetview"), for: .normal)
        restartTrackingBtn.tintColor = UIColor.gray
    }

    // MARK: - Helpers
    
    func urlFieldText() -> String? {
        return urlField.text
    }
    
    // MARK: - Actions
    
    func startLoading(_ url: String?) {
        permissionLevelButton?.startAnimating()
        cancelBtn?.isHidden = false
        reloadBtn?.isHidden = true
        urlField.text = url
    }
    
    func finishLoading(_ url: String?) {
        permissionLevelButton?.stopAnimating()
        cancelBtn?.isHidden = true
        reloadBtn?.isHidden = false
    }
    
    func setBackEnabled(_ enabled: Bool) {
        backBtn.isEnabled = enabled
    }
    
    func setForwardEnabled(_ enabled: Bool) {
        forwardBtn.isEnabled = enabled
    }
    
    func setDebugSelected(_ selected: Bool) {
        debugBtn.isSelected = selected
    }
    
    func setDebugVisible(_ visible: Bool) {
        debugBtn.isHidden = !visible
    }
    
    func setRestartTrackingVisible(_ visible: Bool) {
        restartTrackingBtn.isHidden = !visible
    }
    
    func hideKeyboard() {
        urlField.resignFirstResponder()
    }
    
    func isDebugButtonSelected() -> Bool {
        return debugBtn.isSelected
    }
    
    func hideCameraFlipButton() {
        switchCameraBtn.removeFromSuperview()
    }
    
    // MARK: - Button Actions
    
    @IBAction func backAction(_ sender: Any) {
        appDelegate().logger.debug("backAction")
        urlField.resignFirstResponder()
        backActionBlock?(sender)
    }

    @IBAction func forwardAction(_ sender: Any) {
        appDelegate().logger.debug("forwardAction")
        urlField.resignFirstResponder()
        forwardActionBlock?(sender)
    }

    @IBAction func homeAction(_ sender: Any) {
        appDelegate().logger.debug("homeAction")
        homeActionBlock?(sender)
    }

    @IBAction func reloadAction(_ sender: Any) {
        appDelegate().logger.debug("reloadAction")
        urlField.resignFirstResponder()
        reloadActionBlock?(sender)
    }

    @IBAction func cancelAction(_ sender: Any) {
        appDelegate().logger.debug("cancelAction")
        urlField.resignFirstResponder()
        cancelActionBlock?(sender)
    }
    
    @IBAction func showPermissionsAction(_ sender: Any) {
        appDelegate().logger.debug("showPermissionsAction")
        urlField.resignFirstResponder()
        showPermissionsActionBlock?(sender)
    }

    @IBAction func debugAction(_ sender: Any) {
        debugBtn.isSelected = !debugBtn.isSelected
        debugButtonToggledAction?(debugBtn.isSelected)
    }

    @IBAction func settingsAction() {
        settingsActionBlock?()
    }

    @IBAction func restartTrackingAction(_ sender: Any) {
        restartTrackingActionBlock?()
    }

    @IBAction func switchCameraAction(_ sender: Any) {
        switchCameraActionBlock?()
    }

    // MARK: - UITextField Delegate

    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        goActionBlock?(textField.text)
        return true
    }

    // MARK: - UIView

    // This function increases the hitboxes of the forwardBtn/backBtn
    
    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        let minXPos = backBtn.frame.maxX
        let maxXPos = forwardBtn.frame.minX

        let increaseValue: CGFloat = (maxXPos - minXPos) / 2

        let icreasedBackRect = CGRect(x: backBtn.frame.origin.x - increaseValue, y: backBtn.frame.origin.y - increaseValue, width: backBtn.frame.size.width + increaseValue * 2, height: backBtn.frame.size.height + increaseValue * 2)

        let icreasedForwardRect = CGRect(x: forwardBtn.frame.origin.x - increaseValue, y: forwardBtn.frame.origin.y - increaseValue, width: forwardBtn.frame.size.width + increaseValue * 2, height: forwardBtn.frame.size.height + increaseValue * 2)

        if icreasedBackRect.contains(point) {
            return backBtn
        }

        if icreasedForwardRect.contains(point) {
            return forwardBtn
        }

        return super.hitTest(point, with: event)
    }
}

class URLTextField: UITextField {
    override func leftViewRect(forBounds bounds: CGRect) -> CGRect {
        return CGRect(x: 0, y: 0, width: CGFloat(URL_FIELD_HEIGHT), height: CGFloat(URL_FIELD_HEIGHT))
    }
}
